### PR TITLE
remote-client support for images

### PR DESCRIPTION
**Pull request by baude**
Signed-off-by: baude <bbaude@redhat.com>

### CODE REVIEW

**Comment by baude** [comment]
@rhatdan @mheon do you guys mind giving this a quick review ... we can hold merge off until the release cut but i need to transition to the tag function right after this so testing can be enabled.

**Comment by rhatdan** [comment]
LGTM, Tests are failing?

**Comment by mheon** [comment]
Lot smaller than I was expecting. I'll do a more thorough review tomorrow morning, but direction seems fine.

File: libpod/adapter/images_remote.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+// +build remoteclient
+
+package adapter
+
+import (
+	"github.com/containers/libpod/libpod"
+)
+
+// Info returns information for the host system and its components
+func (r RemoteRuntime) Images() ([]libpod.InfoData, error) {
```
Comment (vrothberg) [inline libpod/adapter/images_remote.go]: According to the comment, the func should be renamed to `Info()`.

File: libpod/adapter/runtime_remote.go [hunk 2]
```diff
@@ -2,27 +2,174 @@
...
+	"github.com/containers/libpod/libpod/image"
+	"github.com/opencontainers/go-digest"
+	"github.com/urfave/cli"
+	"github.com/varlink/go/varlink"
+	"strings"
+	"time"
+)
+
+// ImageRuntime is wrapper for image runtime
+type RemoteImageRuntime struct{}
```
Comment (vrothberg) [inline libpod/adapter/runtime_remote.go]: This doesn't seem to be used (yet?).

File: libpod/image/parts.go [hunk 3]
```diff
@@ -7,12 +7,12 @@ import (
...
-// imageParts describes the parts of an image's name
-type imageParts struct {
+// Parts describes the parts of an image's name
+type Parts struct {
 	transport   string
-	registry    string
+	Registry    string
 	name        string
-	tag         string
+	Tag         string
```
Comment (vrothberg) [inline libpod/image/parts.go]: nitpick: could we move public members to the top and private ones down? The public ones need comments.

File: libpod/adapter/runtime_remote.go [hunk 4]
```diff
@@ -2,27 +2,174 @@
...
+func (ci *ContainerImage) Names() []string {
+	return ci.remoteImage.Names
+}
+
+// Created returns the time the image was created
+func (ci *ContainerImage) Created() time.Time {
+	return ci.remoteImage.Created
+}
+
+// Size...
```
Comment (vrothberg) [inline libpod/adapter/runtime_remote.go]: I think we can be more verbose than that :^)

File: libpod/adapter/runtime_remote.go [hunk 5]
```diff
@@ -2,27 +2,174 @@
...
+	usize := uint64(ci.remoteImage.Size)
+	return &usize, nil
+}
+
+// Digest returns the image's digest
+func (ci *ContainerImage) Digest() digest.Digest {
+	return ci.remoteImage.Digest
+}
+
+// Labels...
```
Comment (vrothberg) [inline libpod/adapter/runtime_remote.go]: I think we can be more verbose than that :^)

File: cmd/podman/imagefilters/filters.go [hunk 6]
```diff
@@ -1,48 +1,49 @@
-package image
+package imagefilters
```
[the remark this answers is not in the document]
Comment (baude) [reply cmd/podman/imagefilters/filters.go]: circular deps ... and it is only used in main

File: cmd/podman/imagefilters/filters.go [hunk 7]
```diff
@@ -1,48 +1,49 @@
...
 
 import (
 	"context"
+	"github.com/containers/libpod/libpod/adapter"
```
[the remark this answers is not in the document]
Comment (baude) [reply cmd/podman/imagefilters/filters.go]: sure, as soon as we figure out how it will be wired :)

File: libpod/adapter/images_remote.go (hunk 1 above)
Comment (baude) [reply libpod/adapter/images_remote.go]: fixed

File: libpod/adapter/runtime_remote.go (hunk 4 above)
Comment (baude) [reply libpod/adapter/runtime_remote.go]: this code is going to be deprecated and removed very very soon

File: libpod/image/parts.go (hunk 3 above)
Comment (baude) [reply libpod/image/parts.go]: this code is going to be deprecated and removed very very soon

File: libpod/adapter/runtime_remote.go (hunk 2 above)
Comment (baude) [reply libpod/adapter/runtime_remote.go]: correct

**Comment by mheon** [comment]
LGTM

File: cmd/podman/images.go [hunk 8]
```diff
@@ -145,20 +145,20 @@ var (
...
-		newImage    *image.Image
+		filterFuncs []imagefilters.ResultFilter
+		newImage    *adapter.ContainerImage
 	)
 	if err := validateFlags(c, imagesFlags); err != nil {
 		return err
 	}
 
-	runtime, err := libpodruntime.GetRuntime(c)
+	localRuntime, err := adapter.GetRuntime(c)
```
Comment (mtrmac) [inline cmd/podman/images.go]: FWIW I find it rather confusing that the `runtime` object, which may either be an actual local state, or a RPC client to a _remote_ server, is named `localRuntime` — both here, and the `runtime_remote.go:LocalRuntime` struct.

I would expect something like a `Runtime interface{…}`, with a `[Ll]ocalRuntime` and `[Rr]emoteRuntime` implementations — which could even be both compiled in at the same time, on platforms where that make sense, and allow using the same binary either for local or remote operations, depending on CLI options.

File: libpod/adapter/images_remote.go [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+)
+
+// Images returns information for the host system and its components
+func (r RemoteRuntime) Images() ([]libpod.InfoData, error) {
+	conn, err := r.Connect()
+	if err != nil {
+		return nil, err
+	}
+	_ = conn
+	return nil, nil
```
Comment (mtrmac) [inline libpod/adapter/images_remote.go]: ??

File: cmd/podman/imagefilters/filters.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (vrothberg) [reply cmd/podman/imagefilters/filters.go]: I don't think that would cause a circular dependency as it is another package.

**Pull request merged**